common/cpp/src/google_smart_card_common/value_conversion.h

```diff
@@ -27,21 +27,153 @@
+// value, by finding the corresponding item among visited ones. This is a
+// non-template class, because enum values are casted to integers.
+class EnumFromValueConverter final {
+ public:
+  explicit EnumFromValueConverter(Value value_to_convert);
+  EnumFromValueConverter(const EnumFromValueConverter&) = delete;
+  EnumFromValueConverter& operator=(const EnumFromValueConverter&) = delete;
+  ~EnumFromValueConverter();
+
+  void HandleItem(int64_t enum_item, const char* enum_item_name);
+  bool GetConvertedEnum(const char* type_name, int64_t* converted_enum,
+                        std::string* error_message) const;
+
+ private:
+  const Value value_to_convert_;
+  optional<int64_t> converted_enum_;
+};
+
+}  // namespace internal
+
+///////////// EnumValueDescriptor ////////////
+
+// Class that allows to describe an enum's items and their corresponding string
+// names. Can be used in order to automatically implement conversion of a C++
+// enum to/from a string `Value` object (provided by the global template
+// functions `ConvertToValue()` and `ConvertFromValue()` defined below).
+//
+// Example usage:
+//   enum class Tp { kA, kB };
+//   template <> EnumValueDescriptor<Tp>::Description
+//   EnumValueDescriptor<Tp>::GetDescription() {
+//     return Describe("Tp").WithItem(Tp::kA, "a").WithItem(Tp::kB, "b");
+//   }
+//   ...
+//   ConvertToValue(Tp::kA, ...);
+//   ConvertFromValue(Value("b"), ...);
+template <typename T>
+class EnumValueDescriptor {
+ public:
+  static_assert(std::is_enum<T>::value,
+                "The EnumValueDescriptor parameter must be an enum");
+
+  // Class that should be used for describing items of the enum type. Should be
+  // instantiated via the `Describe()` method.
+  class Description final {
+   public:
+    Description(const char* type_name,
+                internal::EnumToValueConverter* to_value_converter,
+                internal::EnumFromValueConverter* from_value_converter)
+        : type_name_(type_name),
+          to_value_converter_(to_value_converter),
+          from_value_converter_(from_value_converter) {}
+    Description(Description&&) = default;
+    Description& operator=(Description&&) = default;
+    ~Description() = default;
+
+    const char* type_name() const { return type_name_; }
+
+    // Adds the given item into the enum's description: |enum_item_name| is the
+    // `Value` representation of |enum_item|.
+    // Returns a rvalue reference to |this| and uses the "&&" ref-qualifier, so
+    // that the method calls can be easily chained and the final result can be
+    // returned without an explicit std::move() boilerplate.
+    Description&& WithItem(T enum_item, const char* enum_item_name) && {
+      const int64_t enum_item_number = static_cast<int64_t>(enum_item);
+      if (to_value_converter_)
+        to_value_converter_->HandleItem(enum_item_number, enum_item_name);
+      else
+        from_value_converter_->HandleItem(enum_item_number, enum_item_name);
```

Comment:
what if from_value_converter is nullptr?

Reply:
Done - Even though exactly one of pointers should be non-null, it costs almost nothing to check it here explicitly.